File: 2024/haskell/day01a/Main.hs
module Main ( main ) where

import Data.List ( sort )
import System.Environment ( getArgs, getProgName )
import System.Exit ( exitFailure )

usage :: IO ()
usage = do
  progname <- getProgName
  putStrLn $ "usage: " ++ progname ++ " <file>"
  exitFailure

buildLists :: String -> ([Int], [Int])
buildLists contents =
    foldl
        (\(firstList, secondList) line ->
            let numbersList = words line
                firstNumber = read $ head numbersList :: Int
                secondNumber = read (numbersList !! 1) :: Int
            in (firstList ++ [firstNumber], secondList ++ [secondNumber])
        )
        ([], [])
        (lines contents)

computeDistances :: [(Int, Int)] -> [Int]
computeDistances = map (\(first, second) -> abs (first - second))

process :: String -> Int
process contents =
    let (firstList, secondList) = buildLists contents
        sortedFirst = sort firstList
        sortedSecond = sort secondList
        distances = computeDistances $ zip sortedFirst sortedSecond
    in sum distances

main :: IO ()
main = do
  args <- getArgs
  case args of
    [filename] -> do
      contents <- readFile filename
      let result = process contents
      putStrLn $ "result = " ++ show result
    _ -> usage

File: 2024/rust/day01a/src/main.rs
use std::env;
use std::fs;
use std::process;

fn usage() {
    let progname = env::args().next().unwrap();
    println!("usage: {progname} <file>");
    process::exit(1);
}

fn build_lists(contents: &str) -> (Vec<u32>, Vec<u32>) {
    let mut first_list = vec![];
    let mut second_list = vec![];
    for line in contents.lines() {
        let first_number = line
            .split_whitespace()
            .next()
            .unwrap()
            .parse::<u32>()
            .unwrap();
        let second_number = line
            .split_whitespace()
            .nth(1)
            .unwrap()
            .parse::<u32>()
            .unwrap();
        first_list.push(first_number);
        second_list.push(second_number);
    }
    first_list.sort();
    second_list.sort();
    (first_list, second_list)
}

fn process(contents: &str) -> u32 {
    let mut result = 0;
    let (first_list, second_list) = build_lists(contents);
    for (first, second) in first_list.iter().zip(second_list.iter()) {
        result += first.abs_diff(*second);
    }
    result
}

fn main() {
    if env::args().count() < 2 {
        usage();
    }
    let filename = env::args().nth(1).unwrap();
    let contents = fs::read_to_string(filename).expect("read of input file failed");
    let result = process(&contents);
    println!("result = {result}");
}
